File: benchmarks/pandas/bench_natsort.py
"""Benchmark: natsort — natural-order sorting of 10k strings with numeric suffixes.

Mirrors tsb's natSorted / natCompare / natSortKey / natArgSort using the
Python `natsort` package (falls back to a manual key if natsort not installed).
"""
import json, time

N = 10_000
WARMUP = 3
ITERATIONS = 10

# Build the same dataset as the TS benchmark
items = [f"item{N - i}" for i in range(N)]

try:
    from natsort import natsorted, natsort_keygen
    nat_key = natsort_keygen()
    def run():
        natsorted(items)
        nat_key("file42")
except ImportError:
    # Fallback: manual digit-aware key (equivalent logic)
    import re
    def _nat_key(s):
        return [int(t) if t.isdigit() else t for t in re.split(r"(\d+)", s)]
    def run():
        sorted(items, key=_nat_key)
        _nat_key("file42")

for _ in range(WARMUP):
    run()

start = time.perf_counter()
for _ in range(ITERATIONS):
    run()
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "natsort",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))

File: benchmarks/tsb/bench_natsort.ts
/**
 * Benchmark: natSorted, natCompare, natSortKey, natArgSort on 10k strings
 *
 * Mirrors Python `natsort` package usage: natural-order sorting of strings
 * with embedded numeric tokens (e.g. "file10" sorts after "file9").
 */
import { natSorted, natCompare, natSortKey, natArgSort } from "../../src/index.js";

const N = 10_000;
const WARMUP = 3;
const ITERATIONS = 10;

// Build an array of strings with numeric suffixes (out of natural order)
const items = Array.from({ length: N }, (_, i) => `item${N - i}`);

// Warm-up
for (let i = 0; i < WARMUP; i++) {
  natSorted(items);
  natCompare("file10", "file9");
  natSortKey("file42");
  natArgSort(items);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  natSorted(items);
  natCompare("file10", "file9");
  natSortKey("file42");
  natArgSort(items);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "natsort",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
